text/0000-try-trait-v2.md

```diff
@@ -0,0 +1,1172 @@
+```rust
+use std::num::NonZeroI32;
+pub struct ResultCodeResidual(NonZeroI32);
+```
+
+With that, it's straight-forward to implement the traits.  `NonZeroI32`'s constructor even does exactly the check we need in `Try::branch`:
+```rust
+impl Try for ResultCode {
+    type Output = ();
+    type Residual = ResultCodeResidual;
+    fn branch(self) -> ControlFlow<Self::Residual> {
+        match NonZeroI32::new(self.0) {
+            Some(r) => ControlFlow::Break(ResultCodeResidual(r)),
+            None => ControlFlow::Continue(()),
+        }
+    }
+    fn from_output((): ()) -> Self {
+        ResultCode::SUCCESS
+    }
+}
+
+impl FromResidual for ResultCode {
+    fn from_residual(r: ResultCodeResidual) -> Self {
+        ResultCode(r.0.into())
+    }
+}
+```
+
+Aside: As a nice bonus, the use of a `NonZero` type in the residual means that `<ResultCode as Try>::branch` [compiles down to a nop](https://rust.godbolt.org/z/GxeYax) on the current nightly.  Thanks, enum layout optimizations!
+
+Now, this is all great for keeping the interface that the other unmigrated C code expects, and can even work in `no_std` if we want.  But it might also be nice to give other *Rust* code that uses it the option to convert things into a `Result` with a more detailed error.
+
+For expository purposes, we'll use this error type:
+```rust
+#[derive(Debug, Clone)]
+pub struct FancyError(String);
+```
+
+(A real one would probably be more complicated and have a better name, but this will work for what we need here -- it's bigger and needs non-core things to work.)
+
+We can allow `?` on a `ResultCode` in a method returning `Result` with an implementation like this:
+```rust
+impl<T, E: From<FancyError>> FromResidual<ResultCodeResidual> for Result<T, E> {
+    fn from_residual(r: ResultCodeResidual) -> Self {
+        Err(FancyError(format!("Something fancy about {} at {:?}", r.0, std::time::SystemTime::now())).into())
+    }
+}
+```
+
+*The split between different error strategies in this section is inspired by [`windows-rs`](https://github.com/microsoft/windows-rs), which has both [`ErrorCode`](https://microsoft.github.io/windows-docs-rs/doc/bindings/windows/struct.ErrorCode.html) -- a simple newtype over `u32` -- and [`Error`](https://microsoft.github.io/windows-docs-rs/doc/bindings/windows/struct.Error.html) -- a richer type that can capture a stack trace, has an `Error` trait implementation, and can carry additional debugging information -- where the former can be converted into the latter.*
+
+
+## Using these traits in generic code
+
+`Iterator::try_fold` has been stable to call (but not to implement) for a while now.  To illustrate the flow through the traits in this RFC, lets implement our own version.
```

Comment:
Extremely minor typo:
```suggestion
`Iterator::try_fold` has been stable to call (but not to implement) for a while now.  To illustrate the flow through the traits in this RFC, let's implement our own version.
```

Reply:
Thanks!  Apparently I noticed that when I was making https://github.com/rust-lang/rust/pull/84092/files#diff-55ca2bad04d574c9b9d2b0ea41dc8e3825970da85fd2fd853446e01715f1c024R10 but forgot to fix it here 😅